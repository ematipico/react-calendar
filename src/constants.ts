// https://www.w3.org/TR/uievents-key/#named-key-attribute-values
export const KEY_NAMES = {
	ARROW_RIGHT: 'ArrowRight',
	ARROW_LEFT: 'ArrowLeft',
	ARROW_UP: 'ArrowUp',
	ARROW_DOWN: 'ArrowDown',
};

export const Formats = {
	EEEE_MM_YYYY: 'EEEE, MMM yyyy',
	MMMM: 'MMMM'
};

