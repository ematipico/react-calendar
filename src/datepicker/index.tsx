import React from "react";
import { DatePickerProvider } from "./DatePickerProvider";
import { DatePickerWrapper } from "./DatePickerWrapper";
import { DatePickerHead } from "./DatePickerHead";
import { DatePickerBody } from "./DatePickerBody";

export enum DatePickerViews {
	Days = "Days",
	Months = "Months",
	Years = "Years"
}

export const Days = {
	Sunday: 0,
	Monday: 1,
	Tuesday: 2,
	Wednesday: 3,
	Thursday: 4,
	Friday: 5,
	Saturday: 6
};

export type ShouldDisabled = (currentDate: Date) => boolean;
export type OnDateChosen = (date: Date) => void;


export interface DatePickerProps {

	/**
	 * Called when a date is chosen
	 */
	onDateChosen?: OnDateChosen;

	/**
	 *
	 */
	minDate?: Date;

	/**
	 *
	 */
	maxDate?: Date;

	/**
	 *
	 */
	currentDate?: Date;

	/**
	 *
	 */

	currentView?: DatePickerViews;

	/**
	 *
	 */
	weekStartsOn?: 0 | 1 | 2 | 3 | 4 | 5 | 6;

	/**
	 * An array of dates to disable
	 */
	disabledDates?: Date[]

	/**
	 * It marks a specific cell month disabled or not.
	 */
	shouldDisableMonth?: ShouldDisabled;
	/**
	 * It marks a specific cell month disabled or not.
	 */
	shouldDisableYear?: ShouldDisabled;
	/**
	 * It marks a specific cell month disabled or not.
	 */
	shouldDisableDay?: ShouldDisabled;

	/**
	 * The list of views to support. By default it supports all of them: day, year and month
	 */
	views?: DatePickerViews[];
}

export default function DatePicker(props: DatePickerProps) {
	return (
		<DatePickerProvider>
			<DatePickerWrapper>
				<DatePickerHead />
				<DatePickerBody />
			</DatePickerWrapper>
		</DatePickerProvider>
	);
}
