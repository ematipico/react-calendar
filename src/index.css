@import "../node_modules/normalize.css/normalize.css";
@import "./datepicker/DatePickerWrapper.css";
@import "./datepicker/DatePickerBody.css";
@import "./datepicker/DatePickerHead.css";
@import "./datepicker/DatePickerCell.css";
