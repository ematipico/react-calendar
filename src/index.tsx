import DatePicker from './datepicker/index';

export { DatePicker };
