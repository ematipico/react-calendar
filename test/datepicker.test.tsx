import { render, fireEvent } from '@testing-library/react';
import * as React from 'react';
import { DatePicker } from '../src';
import { format } from 'date-fns';

describe('Datepicker', () => {
	it('should render the day view by default', () => {
		const { queryByText } = render(<DatePicker />);

		const month = format(new Date(), 'MMMM');

		expect(queryByText(month)).toBeTruthy();
	});

	it('should render when a current date is passed', () => {
		const date = new Date(2018, 11, 11);
		const { getByText, queryByText, container } = render(<DatePicker currentDate={date} />);

		expect(queryByText('December')).toBeTruthy();
		const control = getByText('December');
		expect(container.firstChild).toMatchSnapshot();

		fireEvent.click(control);
		expect(queryByText('2010 - 2029')).toBeTruthy();
	});

	it('should correctly render the control label', () => {
		const date = new Date(2019, 11, 11);
		const { getByText, queryByText } = render(<DatePicker currentDate={date} />);

		expect(queryByText('December')).toBeTruthy();
		let control = getByText('December');
		fireEvent.click(control);
		expect(queryByText('2010 - 2029')).toBeTruthy();
		control = getByText('2010 - 2029');
		fireEvent.click(control);
		expect(queryByText('2019')).toBeTruthy();
	});
});
