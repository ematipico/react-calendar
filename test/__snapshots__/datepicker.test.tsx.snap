// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datepicker should render when a current date is passed 1`] = `
<div
  class="DatePicker"
>
  <header
    class="DatePicker__Head"
  >
    <div
      class="DatePicker__Controls"
    >
      <button
        aria-label="Go to the next view"
        class="DatePicker__Control DatePicker__ViewControl"
        type="button"
      >
        December
      </button>
      <div>
        <button
          aria-label="Go to the previous month"
          class="DatePicker__Control DatePicker__Arrows DatePicker__Previous"
          type="button"
        >
          <svg
            class="feather feather-chevron-left"
            fill="none"
            height="32"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="15 18 9 12 15 6"
            />
          </svg>
        </button>
        <button
          aria-label="Go to the next month"
          class="DatePicker__Control DatePicker__Arrows DatePicker__Next"
          type="button"
        >
          <svg
            class="feather feather-chevron-right"
            fill="none"
            height="32"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 18 15 12 9 6"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="DatePicker__Days"
    >
      <abbr
        aria-label="Sunday"
        class="DatePicker__DayName"
      >
        Sun
      </abbr>
      <abbr
        aria-label="Monday"
        class="DatePicker__DayName"
      >
        Mon
      </abbr>
      <abbr
        aria-label="Tuesday"
        class="DatePicker__DayName"
      >
        Tue
      </abbr>
      <abbr
        aria-label="Wednesday"
        class="DatePicker__DayName"
      >
        Wed
      </abbr>
      <abbr
        aria-label="Thursday"
        class="DatePicker__DayName"
      >
        Thu
      </abbr>
      <abbr
        aria-label="Friday"
        class="DatePicker__DayName"
      >
        Fri
      </abbr>
      <abbr
        aria-label="Saturday"
        class="DatePicker__DayName"
      >
        Sat
      </abbr>
    </div>
  </header>
  <main
    class="DatePicker__Body"
    role="grid"
  >
    <div
      class="DatePicker__Row"
    >
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Monday, Nov 2018"
          class="DatePicker__Cell DatePicker__Cell--disabled"
          disabled=""
        >
          26
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Tuesday, Nov 2018"
          class="DatePicker__Cell DatePicker__Cell--disabled"
          disabled=""
        >
          27
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Wednesday, Nov 2018"
          class="DatePicker__Cell DatePicker__Cell--disabled"
          disabled=""
        >
          28
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Thursday, Nov 2018"
          class="DatePicker__Cell DatePicker__Cell--disabled"
          disabled=""
        >
          29
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Friday, Nov 2018"
          class="DatePicker__Cell DatePicker__Cell--disabled"
          disabled=""
        >
          30
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Saturday, Dec 2018"
          class="DatePicker__Cell"
        >
          1
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Sunday, Dec 2018"
          class="DatePicker__Cell"
        >
          2
        </button>
      </div>
    </div>
    <div
      class="DatePicker__Row"
    >
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Monday, Dec 2018"
          class="DatePicker__Cell"
        >
          3
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Tuesday, Dec 2018"
          class="DatePicker__Cell"
        >
          4
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Wednesday, Dec 2018"
          class="DatePicker__Cell"
        >
          5
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Thursday, Dec 2018"
          class="DatePicker__Cell"
        >
          6
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Friday, Dec 2018"
          class="DatePicker__Cell"
        >
          7
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Saturday, Dec 2018"
          class="DatePicker__Cell"
        >
          8
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Sunday, Dec 2018"
          class="DatePicker__Cell"
        >
          9
        </button>
      </div>
    </div>
    <div
      class="DatePicker__Row"
    >
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Monday, Dec 2018"
          class="DatePicker__Cell"
        >
          10
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Tuesday, Dec 2018"
          class="DatePicker__Cell"
        >
          11
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Wednesday, Dec 2018"
          class="DatePicker__Cell"
        >
          12
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Thursday, Dec 2018"
          class="DatePicker__Cell"
        >
          13
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Friday, Dec 2018"
          class="DatePicker__Cell"
        >
          14
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Saturday, Dec 2018"
          class="DatePicker__Cell"
        >
          15
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Sunday, Dec 2018"
          class="DatePicker__Cell"
        >
          16
        </button>
      </div>
    </div>
    <div
      class="DatePicker__Row"
    >
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Monday, Dec 2018"
          class="DatePicker__Cell"
        >
          17
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Tuesday, Dec 2018"
          class="DatePicker__Cell"
        >
          18
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Wednesday, Dec 2018"
          class="DatePicker__Cell"
        >
          19
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Thursday, Dec 2018"
          class="DatePicker__Cell"
        >
          20
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Friday, Dec 2018"
          class="DatePicker__Cell"
        >
          21
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Saturday, Dec 2018"
          class="DatePicker__Cell"
        >
          22
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Sunday, Dec 2018"
          class="DatePicker__Cell"
        >
          23
        </button>
      </div>
    </div>
    <div
      class="DatePicker__Row"
    >
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Monday, Dec 2018"
          class="DatePicker__Cell"
        >
          24
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Tuesday, Dec 2018"
          class="DatePicker__Cell"
        >
          25
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Wednesday, Dec 2018"
          class="DatePicker__Cell"
        >
          26
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Thursday, Dec 2018"
          class="DatePicker__Cell"
        >
          27
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Friday, Dec 2018"
          class="DatePicker__Cell"
        >
          28
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Saturday, Dec 2018"
          class="DatePicker__Cell"
        >
          29
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Sunday, Dec 2018"
          class="DatePicker__Cell"
        >
          30
        </button>
      </div>
    </div>
    <div
      class="DatePicker__Row"
    >
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Monday, Dec 2018"
          class="DatePicker__Cell"
        >
          31
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Tuesday, Jan 2019"
          class="DatePicker__Cell DatePicker__Cell--disabled"
          disabled=""
        >
          1
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Wednesday, Jan 2019"
          class="DatePicker__Cell DatePicker__Cell--disabled"
          disabled=""
        >
          2
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Thursday, Jan 2019"
          class="DatePicker__Cell DatePicker__Cell--disabled"
          disabled=""
        >
          3
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Friday, Jan 2019"
          class="DatePicker__Cell DatePicker__Cell--disabled"
          disabled=""
        >
          4
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Saturday, Jan 2019"
          class="DatePicker__Cell DatePicker__Cell--disabled"
          disabled=""
        >
          5
        </button>
      </div>
      <div
        class="DatePicker__CellWrapper"
        role="gridcell"
      >
        <button
          aria-label="Sunday, Jan 2019"
          class="DatePicker__Cell DatePicker__Cell--disabled"
          disabled=""
        >
          6
        </button>
      </div>
    </div>
  </main>
</div>
`;
